### PR TITLE
Refactor Navigation Cube

### LINKED ISSUE #6818: [Feature Request] Add an option to lock the NaviCube
### Is there an existing request for this?

- [X] I have searched the existing requests

### Forums discussion

https://forum.freecadweb.org/viewtopic.php?f=34&t=68390

### Subproject(s) affected?

_No response_

### Idea description

Sometimes it happens that the NaviCube is dragged by accident. To avoid this an option to lock it would be nice.

In the preferences dialog a check box can be added that affects all existing and newly created NaviCubes. Then in the context-menu there could be this option too but only affects this instance.

### Anything else?

_No response_

### Code of Conduct

- [X] I agree to follow this project's Code of Conduct

### LINKED ISSUE #6935: [Bug] navicube location sometimes jumps to other side
### Is there an existing issue for this?

- [X] I have searched the existing issues

### Forums discussion

https://forum.freecadweb.org/viewtopic.php?f=10&t=69003

### Version

0.19 (Release)

### Full version info

```shell
OS: Windows 10 Version 2009
Word size of OS: 64-bit
Word size of FreeCAD: 64-bit
Version: 0.19.24267 +148 (Git)
Build type: Release
Branch: Branch_0.19.4
Hash: 476ecf091941bead59b14e44afa6064d5a66afa3
Python version: 3.8.6+
Qt version: 5.15.2
Coin version: 4.0.1
OCC version: 7.5.3
Locale: German/Germany (de_DE)
```

### Issue description

- set the NaviCube location to TopLeft
- grab the right border of the main window and shrink the window it to the minimum
result: The NaviCube jumps to the right edge and stays there:
FreeCAD_jD61WCSdcb

### LINKED ISSUE #9262: [Issue] Navigation cube font size strange behaviour
### Is there an existing issue for this?

- [X] I have searched the existing issues

### Forums discussion

https://forum.freecad.org/viewtopic.php?p=674532#p674532

### Version

0.21 (Development)

### Full version info

```shell
[code]
OS: macOS 10.16
Word size of FreeCAD: 64-bit
Version: 0.21.0.31917 (Git)
Build type: Release
Branch: master
Hash: 4639283b459f9ff9093e289908cc003db8745b82
Python 3.10.9, Qt 5.15.6, Coin 4.0.0, Vtk 9.1.0, OCC 7.6.3
Locale: C/Default (C)
Installed mods: 
  * kugghjul.py
  * dxf-library
  * fcgear 1.0.0
  * __pycache__
  * sheetmetal 0.2.61
  * tida.py
  * Curves 0.6.8
  * Assembly4 0.12.7
[/code]
```

### Subproject(s) affected?

Core

### Issue description

• The font size setting of the navigation cube does not match the size of the navigation cube. Setting the font size to 100 gives a 4 times smaller font. Image showing cube size 250 and font size 250
Skärmavbild 2023-04-13 kl  19 12 08
• Setting a too big font size makes the text beed over the edges of the face it should lie on.
Image showing cube size 250 and font size 900
Skärmavbild 2023-04-13 kl  19 08 02
• If tuning the font size to match the cube size and then change the cube size you have to tune the font size again. 

### Anything else?

_No response_

### Code of Conduct

- [X] I agree to follow this project's Code of Conduct

### PR DESCRIPTION
**Pull request by Jolbas**
This PR does not conform to the one issue recommendation. Hopefully you can bear reviewing it anyway.
Fixes #9262 
Fixes #6935 
Fixed #6818 
But also a non registered issue from the forum:
[https://forum.freecad.org/viewtopic.php?t=77768](url)
Will hopefully also fix a problem on Raspberry Pi
[https://forum.freecad.org/viewtopic.php?t=76438](url)

Here is a list of changes
- Uses std commands in menu coming with icons
- Cube only draggable if activated in menu 
- Skärmavbild 2023-04-22 kl  17 14 49
- Ortographic and Perspective reflected in Navigation Cube
- Skärmavbild 2023-04-22 kl  17 18 21
- Skärmavbild 2023-04-22 kl  17 18 32
- Remove textures except for the labels and use gl fill instead
- Borders on surrounding buttons to make better contrast.
- Fix some openGL issues
- Remove duplicate parameter management. Kept and improved the one in View3DSettings
- Remove lots of other redundant code
- Replaced FontSize param with FontZoom and enabled auto zoom/size on labels.
   - FontZoom = 1.0 makes all labels as big as possible individually
   - Skärmavbild 2023-04-22 kl  17 20 30
   - FontZoom = 0.0 As big as possible but same font size on all.
   - Skärmavbild 2023-04-22 kl  17 20 54
   - Values in between limits max font size. Default 0.3
   - FontZoom < 0.0 Same font size on all but downscaled.
- Merged ButtonColor and FrontColor to BaseColor parameter
- BaseColor in GUI settings instead of ButtonColor.
- Merged BorderColor and TextColor to single EmphaseColor parameter
- EmphaseColor defaults to black or whit whichever gives most contrast to BaseColor. 
- Skärmavbild 2023-04-22 kl  17 11 39
- Skärmavbild 2023-04-22 kl  17 12 01
- Added ChamferSize parameter to control edge and corner size
- Skärmavbild 2023-04-22 kl  18 04 36
- Skärmavbild 2023-04-22 kl  18 04 55
- Use glDrawArrays() innstead of glBegin() and glEnd(). Better compatibility and faster.
- Draw butt ends on RGB coordinate system
- Fixe issues where cube switch corner on resize window

[X ]  The Pull Request meets the requirements i think.

### CODE REVIEW

**Comment by freecadci** [comment]
<a href="https://gitlab.com/freecad/FreeCAD-CI/-/commits/PR_9356">pipeline status</a> for feature branch [PR_9356](https://gitlab.com/freecad/FreeCAD-CI/-/commits/PR_9356). Pipeline [845397286](https://gitlab.com/freecad/FreeCAD-CI/-/pipelines/845397286) was triggered at [3361508](https://github.com/FreeCAD/FreeCAD/pull/9356/commits/336150880ab3b9f8bd72cd2791b60bd8730b0a67). All CI [branches](https://gitlab.com/freecad/FreeCAD-CI/-/branches/all) and [pipelines](https://gitlab.com/freecad/FreeCAD-CI/-/pipelines?scope=branches).

**Comment by xtemp09** [comment]
@Jolbas, does it mean that the minimal OpenGL version now 1.1?

**Comment by freecadci** [comment]
<a href="https://gitlab.com/freecad/FreeCAD-CI/-/commits/PR_9356">pipeline status</a> for feature branch [PR_9356](https://gitlab.com/freecad/FreeCAD-CI/-/commits/PR_9356). Pipeline [845479779](https://gitlab.com/freecad/FreeCAD-CI/-/pipelines/845479779) was triggered at [ee35112](https://github.com/FreeCAD/FreeCAD/pull/9356/commits/ee35112162a58d4ddcb7625fb8cf50618578c871). All CI [branches](https://gitlab.com/freecad/FreeCAD-CI/-/branches/all) and [pipelines](https://gitlab.com/freecad/FreeCAD-CI/-/pipelines?scope=branches).

**Comment by Rexbas** [comment]
Nice work, I tested it and tried to break something but I did not manage to break it. You also fixed another bug that I have had (but I never reported). I used to have some graphical bugs like in the image below, but only for very particular camera orientations. This does not happen anymore with this PR :)
image
image
I do have two questions. I noticed that the red, blue and green lines no longer intersect in one point. Is this intentional?
image
And how do I modify the ChamferSize parameter. I could not find an option to modify it.

One other question that is not related to this PR but could someone explain to me why there is a `NaviCube` class and a `NaviCubeImplementation` class. Seems to me that `NaviCube` methods are directly forwarded to an equally named method in `NaviCubeImplementation` or they set a member variable in `NaviCubeImplementation`. If there is no good reason for this then I think these two classes can be combined into one `NaviCube` class in the future.

**Comment by Jolbas** [comment]
@Rexbas Some parameters have no GUI and you must use the Menu->Tools->Edit parameters...

Supported parameters with this PR

- CornerNaviCube: 0,1,2 or 3 (1)
- CubeSize: limited in the GUI setting to 1024 (132)
- ChamferSize: 0.05 to 0.20 (0.12)
- NaviRotateToNearest: bool (true)
- FontZoom: -1.0 to 1.0 (0.3)
- FontString: Name of font (depends on platform)
- FontWeight: 0 to 100 where zero means ignore (0)
- FontStretch: 0 to 100 where zero means ignore (0)
- BaseColor: unsigned 0 to 2^32-1 (3806916544)
- EmphaseColor: unsigned 0 to 2^32-1 (black or white depending on BaseColor)
- HiliteColor:: unsigned 0 to 2^32-1 (2867003391)
- OffsetX and OffsetY: Increase margin to window border (0)
- TextFront,TextTop,TextRight,TextRear,TextBottom,TextLeft: string override labels 
- ShowCS:bool (true)
- BorderWidth: 1.0-8.0 or more, may depend on OpenGL driver (1.1)

The gap in rgb coordinate system was intentional. I thought it looked nice but not important to me at all so I'll be happy change it if you don't like it.

I also have wondered about the split in classes. I don't know why but haven't change it. Think it would make more sense to split it in one singleton NavicubeData class that holds the shape and textures that is shared by the NaviCube Instances which only holds the viewer and temporary window position.

**Comment by Rexbas** [comment]
Are all those parameters supposed to be here?
image
I tried to search for ChamferSize but could not find it.

<br><br>

> I thought it looked nice but not important to me at all so I'll be happy change it if you don't like it.

No need to change it, I was just wondering and I personally don´t mind whether there is a small gap or not.

**Comment by Jolbas** [comment]
@Rexbas  If it's not there it falls back to the default. If you want to override you need to add it. Right click in the list and create a new of correct type.

**Comment by Rexbas** [comment]
Ah yes it works. Nice thanks.

File: src/Gui/NaviCube.cpp [hunk 1]
```diff
@@ -131,112 +96,103 @@ class NaviCubeImplementation : public ParameterGrp::ObserverType {
...
-        TEX_ARROW_NORTH,
-        TEX_ARROW_SOUTH,
-        TEX_ARROW_EAST,
-        TEX_ARROW_WEST,
-        TEX_ARROW_RIGHT,
-        TEX_ARROW_LEFT,
-        TEX_DOT_BACKSIDE,
-        TEX_VIEW_MENU_ICON,
-        TEX_VIEW_MENU_FACE
+    enum {
```
Comment (chennes) [inline src/Gui/NaviCube.cpp:118]: You could consider making this an enum class, and dropping the `PID_` prefix -- as a general rule the C++ Core Guidelines encourage the use of enum class wherever possible (but I haven't looked to see if you need to use this as an int or things like that, so this is just an option to look into).

Comment (Jolbas) [reply src/Gui/NaviCube.cpp:118]: Sounds good to me. It is used to color code the surfaces of the cube to find which one is hovered by the pointer but I can cast to int in that case

Comment (chennes) [reply src/Gui/NaviCube.cpp:118]: Do you want to do that as part of this PR, or as a follow-up?

Comment (Jolbas) [reply src/Gui/NaviCube.cpp]: @chennes It's done. Should I squash those commits to one?

**Comment by freecadci** [comment]
<a href="https://gitlab.com/freecad/FreeCAD-CI/-/commits/PR_9356">pipeline status</a> for feature branch [PR_9356](https://gitlab.com/freecad/FreeCAD-CI/-/commits/PR_9356). Pipeline [846053468](https://gitlab.com/freecad/FreeCAD-CI/-/pipelines/846053468) was triggered at [cfe9a8d](https://github.com/FreeCAD/FreeCAD/pull/9356/commits/cfe9a8d529493a2383412776e32874b7bc9f514f). All CI [branches](https://gitlab.com/freecad/FreeCAD-CI/-/branches/all) and [pipelines](https://gitlab.com/freecad/FreeCAD-CI/-/pipelines?scope=branches).

**Comment by freecadci** [comment]
<a href="https://gitlab.com/freecad/FreeCAD-CI/-/commits/PR_9356">pipeline status</a> for feature branch [PR_9356](https://gitlab.com/freecad/FreeCAD-CI/-/commits/PR_9356). Pipeline [846336914](https://gitlab.com/freecad/FreeCAD-CI/-/pipelines/846336914) was triggered at [d0f35c4](https://github.com/FreeCAD/FreeCAD/pull/9356/commits/d0f35c45f3c0889d8586af21f429fcb7e13c4144). All CI [branches](https://gitlab.com/freecad/FreeCAD-CI/-/branches/all) and [pipelines](https://gitlab.com/freecad/FreeCAD-CI/-/pipelines?scope=branches).

File: src/Gui/NaviCube.cpp (hunk 1 above)
Comment (Jolbas) [reply src/Gui/NaviCube.cpp]: Now it passed. There was a problem triggered by the identifier OUT in the Windows build. Seems there is a macro triggered by it, I'm not sure. Not a very good name on a macro in that case? I switched to CamelCase

**Comment by freecadci** [comment]
<a href="https://gitlab.com/freecad/FreeCAD-CI/-/commits/PR_9356">pipeline status</a> for feature branch [PR_9356](https://gitlab.com/freecad/FreeCAD-CI/-/commits/PR_9356). Pipeline [846385313](https://gitlab.com/freecad/FreeCAD-CI/-/pipelines/846385313) was triggered at [466fc32](https://github.com/FreeCAD/FreeCAD/pull/9356/commits/466fc3221159a9775f99de611d18a0015aaf9861). All CI [branches](https://gitlab.com/freecad/FreeCAD-CI/-/branches/all) and [pipelines](https://gitlab.com/freecad/FreeCAD-CI/-/pipelines?scope=branches).

**Comment by adrianinsaval** [comment]
The gap on the coordinate axis looks odd to me, but I don't feel strongly about it

**Comment by chennes** [comment]
Let's merge this one as-is (it's a clear improvement over what we had before), and if lots of people complain about the gap, then we can revisit.

**Comment by Jolbas** [comment]
Great. How about documentation?  Should it be done now and then how or by who?
Wiki page for [Navigation Cube](https://wiki.freecad.org/Navigation_Cube)

**Comment by chennes** [comment]
Now is always better than later 😄  -- if you don't have Wiki editing permissions there is a forums topic you can request them at.

**Comment by chennes** [comment]
Also, don't forget to add the parameters to the "Fine tuning" page on the Wiki

**Comment by adrianinsaval** [comment]
and the release notes

**Comment by wwmayer** [comment]
With this PR I observe a regression. After starting the application then creating a document always results into a main window where the right and bottom part is totally black. In order to manually fix I have to do something that triggers a repaint of the main window (e.g. manually resizing a dock window, creating a second document, interaction with the 3d view, ...).

The issue doesn't occur when closing all documents and creating a new one. So, it's just the first time after starting the application.
Black_area

**Comment by wwmayer** [comment]
Here is a fix for this problem: https://github.com/FreeCAD/FreeCAD/pull/9425

**Comment by wwmayer** [comment]
I have found another more serious regression right now and I don't know if this PR is related.
Whenever you load an image into the 3d view it's rendered completely black. Only after selecting it in the tree view it renders the correct content.

**Comment by Jolbas** [comment]
I check again that all openGL states are reset properly

**Comment by wwmayer** [comment]
This is the change to blame: 8abd6d0870

**Comment by Jolbas** [comment]
@wwmayer I couldn't reproduce the issue using File->Import. How do you load the image?

**Comment by wwmayer** [comment]
File > Open > Select image file or by drag and drop alternatively.

**Comment by Jolbas** [comment]
I could add a `glPushClientAttrib(GL_CLIENT_ALL_ATTRIB_BITS);` in the beginning of `void NaviCubeImplementation::drawNaviCube(bool pickMode)` and a corresponding `glPopClientAttrib` at the end but I can't check if that helps because unfortunately the image is rendered as expected here (nothing black).
```
OS: macOS 12.6
Word size of FreeCAD: 64-bit
Version: 0.21.0.32997 +1 (Git)
Build type: Release
Branch: Navicube_disable_drag
Hash: e3b6c8e4957f65d41788975f80116729ea35e853
Python 3.10.11, Qt 5.15.8, Coin 4.0.0, Vtk 9.2.6, OCC 7.7.0
Locale: English/United States (en_US)
Installed mods: 
  * dxf-library
  * fcgear 1.0.0
  * sheetmetal 0.2.61
  * Curves 0.6.8
  * Assembly4 0.12.7
```

**Comment by wwmayer** [comment]
When using a clean config file the issue doesn't occur. The relevant change to make it happen is to use a single background color (instead of linear or radial gradient).

**Comment by wwmayer** [comment]
> I could add a glPushClientAttrib(GL_CLIENT_ALL_ATTRIB_BITS); in the beginning of

This doesn't help.

**Comment by wwmayer** [comment]
I made an interesting observation:
when the NaviCube is disabled it's still created and in order to make sure it doesn't affect the rendering I modified the code to not even create it. To my big surprise the black image still appears.

Now I went back to commit b3727d907 which is working as expected. When I do the same here to not create the NaviCube then the image becomes black, too.

So, the issue must be caused elsewhere and in the past the NaviCube healed the issue and now it doesn't any more.

**Comment by wwmayer** [comment]
In the meantime I have tested everything on Windows and there everything works normally. So, it's an issue with my Linux system only -- maybe related to the nouveau driver.

**Comment by Jolbas** [comment]
I still can't reproduce with clean config and simple color background. This is a list of GL commands that may have been set in `NaviCubeImplementation::drawNaviCube(bool pickMode)` before the commit but not after the commit.
```
glDepthMask(GL_TRUE);
glBlendFunc(GL_SRC_ALPHA, GL_ONE_MINUS_SRC_ALPHA);
glAlphaFunc(GL_GREATER, 0.25);
glTexParameteri(GL_TEXTURE_2D, GL_TEXTURE_WRAP_S, GL_REPEAT);
glTexParameteri(GL_TEXTURE_2D, GL_TEXTURE_WRAP_T, GL_REPEAT);
glTexParameteri(GL_TEXTURE_2D, GL_TEXTURE_MAG_FILTER, GL_NEAREST);
glTexParameteri(GL_TEXTURE_2D, GL_TEXTURE_MIN_FILTER, GL_NEAREST);
glEnableClientState(GL_TEXTURE_COORD_ARRAY);
```
Could maybe try them all at the end. If the driver doesn't support glPopAttrib() which is the case in OpenGL ES then we may have a lot more to try.

**Comment by wwmayer** [comment]
I tried it but still no luck. I wonder whether it could be related to the changed depth range. Before it was the range [0, 1] while now it's [0.1, 0.9] plus corresponding changes.

**Comment by Jolbas** [comment]
The depth range is and was reset but maybe one of the related states. To night I can go through all states and create a code snippet that mimic previous behaviour.

**Comment by Jolbas** [comment]
Here's a more complete snippet that should mimic previous behaviour.
```
    glDepthMask(GL_TRUE);
    glClearDepth(1.0f);
    glDepthFunc(GL_LEQUAL);

    glDisable(GL_LIGHTING);

    glTexEnvi(GL_TEXTURE_ENV, GL_TEXTURE_ENV_MODE, GL_REPLACE);
    glEnable(GL_TEXTURE_2D);

    glBlendFunc(GL_SRC_ALPHA, GL_ONE_MINUS_SRC_ALPHA);
    glTexEnvf(GL_TEXTURE_ENV, GL_TEXTURE_ENV_MODE, GL_MODULATE);

    glEnable(GL_CULL_FACE);
    glCullFace(GL_BACK);
    glFrontFace(GL_CCW);

    glAlphaFunc(GL_GREATER, 0.25);
    glEnable(GL_ALPHA_TEST);

    glPixelStorei(GL_UNPACK_ALIGNMENT, 1);

    glTexParameteri(GL_TEXTURE_2D, GL_TEXTURE_WRAP_S, GL_REPEAT);
    glTexParameteri(GL_TEXTURE_2D, GL_TEXTURE_WRAP_T, GL_REPEAT);
    if (false) {
        glLineWidth(2.0);
        glDisable(GL_BLEND);
        glTexParameteri(GL_TEXTURE_2D, GL_TEXTURE_MAG_FILTER, GL_NEAREST);
        glTexParameteri(GL_TEXTURE_2D, GL_TEXTURE_MIN_FILTER, GL_NEAREST);
        glShadeModel(GL_FLAT);
        glDisable(GL_DITHER);
        glDisable(GL_POLYGON_SMOOTH);
    }
    else {
    	glShadeModel(GL_SMOOTH);
        glTexParameteri(GL_TEXTURE_2D, GL_TEXTURE_MAG_FILTER, GL_LINEAR);
        glTexParameteri(GL_TEXTURE_2D, GL_TEXTURE_MIN_FILTER, GL_LINEAR);
        glRasterPos3d(-1.1, -1.1, 1.1);
        glLineWidth(m_BorderWidth);
        glPolygonMode(GL_FRONT_AND_BACK, GL_FILL);
    	glEnable(GL_BLEND);
    }
    glDisableClientState(GL_VERTEX_ARRAY);
    glDisableClientState(GL_TEXTURE_COORD_ARRAY);
    glDisable(GL_DEPTH_TEST);
    glClear(GL_DEPTH_BUFFER_BIT);
```
Put it right before the line `// Restore original state.`
Because I'm not absolute sure of the state of pickMode the last time the function is called before your black image is rendered. Try change `if (false) {` to `if (true) {`.

If this doesn't help the problem is probably somewhere else. For example there is a `glLoadIdentity();` without control over glMatrixMode in `NaviCubeImplementation::pickFace(short x, short y)` that previously was called a lot more often. I should by the way remove it when this is sorted out.

**Comment by wwmayer** [comment]
> Here's a more complete snippet that should mimic previous behaviour.

Unfortunately, none of the two tips solves the issue for me. There are a few more findings I have made:
* If after loading an image I use the function Tools > Save image then the output image is correctly displayed while the image in the 3d view remains black
* If I select the image via 3d view or tree view it renders correctly. After de-selecting it the image will become black again
* If I hover over the tree view item for more than a second then the highlighting mechanism is activated and from then on the image is correctly displayed -- independent of the selection status

**Comment by wwmayer** [comment]
Do you know the utility apitrace? After I added some OpenGL debug messages according to https://github.com/apitrace/apitrace/blob/master/docs/USAGE.markdown#opengl-annotations I was able to see the differences between the used OpenGL functions for the case where it works and where it doesn't.

Since the differences are huge and thus still not easy to figure out what exactly the problem is I tried another way. As I already said that when I use a gradient background the problem doesn't occur. So, I checked what are the differences between using a simple and a gradient colored background and here it is:
```
    glMatrixMode(GL_PROJECTION);
    glPushMatrix();
    glLoadIdentity();
    glOrtho(-1, 1, -1, 1, -1, 1);
    glMatrixMode(GL_MODELVIEW);
    glPushMatrix();
    glLoadIdentity();
    glPushAttrib(GL_ENABLE_BIT);
    glDisable(GL_DEPTH_TEST);
    glDisable(GL_LIGHTING);
    glDisable(GL_TEXTURE_2D);
    glBegin(GL_TRIANGLE_STRIP);
    glColor3f(col.redF(), col.greenF(), col.blueF());
    glVertex2f(-1, 1);
    glColor3f(col.redF(), col.greenF(), col.blueF());
    glVertex2f(-1, -1);
    glColor3f(col.redF(), col.greenF(), col.blueF());
    glVertex2f(1, 1);
    glColor3f(col.redF(), col.greenF(), col.blueF());
    glVertex2f(1, -1);
    glEnd();
    glPopAttrib();
    glPopMatrix();
    glMatrixMode(GL_PROJECTION);
    glPopMatrix();
    glMatrixMode(GL_MODELVIEW);
``` 
Important is the part glBegin() ... glEnd() because when I remove it then the image will be black again. Now when I compare the NaviCube code before and after the refactoring then indeed all glBegin() / glEnd() calls have been removed that leads to the problem.

So, I can implement a workaround by adding a function View3DInventorView::drawBackground() that uses the code snippet above.

**Comment by Jolbas** [comment]
Ok. Interesting and strange. I haven't tried apitrace. Will have a look. Is it then enough with the following snippet?
```
glBegin();
glEnd();
```

**Comment by wwmayer** [comment]
> Is it then enough with the following snippet?

No. In between the calls something meaningful must be done.

**Comment by wwmayer** [comment]
https://github.com/FreeCAD/FreeCAD/pull/9450

**Comment by wwmayer** [comment]
Another serious regression is that the Alignment command cannot be used any more: https://forum.freecad.org/viewtopic.php?p=680571#p680571

**Comment by Jolbas** [comment]
Thanks for notify here. I think I know where to look.

**Pull request merged**